Add blog post for Provably Correct Peephole Optimizations with Alive

atucker opened the pull request:
This adds the blog post for Provably Correct Peephole Optimizations with Alive.

atucker committed:
Create index.md

Add the blog post sans the image

atucker committed:
Add files via upload

Add the paper snippet

atucker committed:
Add a bit more 

Add a bit more, including some more critical comments.

atucker committed:
rename folder

atucker committed:
Fix typos in footnote usage and a spelling issue

atucker committed:
Update index.md

Fix front matter

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+link = "https://www.cs.cornell.edu/~aarondtucker/"
++++
+
+## Blog: Provably Correct Peephole Optimizations with Alive
```
Probably no need for "Blog:" in the title here. In fact, probably no need for a title at all—if you look at the rendered output, you'll see that your title gets put up top automatically.

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
++++
+title = "Efficient Path Profiling"
```
This is the title for the other blog post.

atucker replied:
Ah, thanks, my bad!

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+
+## Blog: Provably Correct Peephole Optimizations with Alive
+
+Citation: Nuno P. Lopes, David Menendez, Santosh Nagarakatte, and John Regehr. 2015. Provably correct peephole optimizations with alive. SIGPLAN Not. 50, 6 (June 2015), 22–32. DOI:https://doi.org/10.1145/2813885.2737965
```
TBH I don't think you need to start with a citation, especially since you effectively give a citation in prose right below.

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+Citation: Nuno P. Lopes, David Menendez, Santosh Nagarakatte, and John Regehr. 2015. Provably correct peephole optimizations with alive. SIGPLAN Not. 50, 6 (June 2015), 22–32. DOI:https://doi.org/10.1145/2813885.2737965
+
+## Introduction
+Provably Correct Peephole Optimizations with Alive by Nuno P. Lopes, David Menendez, Santosh Nagarakatte, and John Regehr (SIGPLAN 2015) is a paper with a pretty straightforward goal -- define a domain specific language and semantics for peephole optimizations in LLVM which make it easy to verify the correctness of peephole optimizations, while also building tools which make it simple to actually ...
```
Maybe make a Markdown link to the paper at the beginning there. For extra credit, you can link to the authors' webpages too. :smiley:

atucker replied:
Ooh, good idea!

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+Citation: Nuno P. Lopes, David Menendez, Santosh Nagarakatte, and John Regehr. 2015. Provably correct peephole optimizations with alive. SIGPLAN Not. 50, 6 (June 2015), 22–32. DOI:https://doi.org/10.1145/2813885.2737965
+
+## Introduction
+Provably Correct Peephole Optimizations with Alive by Nuno P. Lopes, David Menendez, Santosh Nagarakatte, and John Regehr (SIGPLAN 2015) is a paper with a pretty straightforward goal -- define a domain specific language and semantics for peephole optimizations in LLVM which make it easy to verify the correctness of peephole optimizations, while also building tools which make it simple to actually ...
```
The conference is PLDI 2015, not "SIGPLAN". (Unfortunately the ACM DL lies about this all the time.)

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+### LLVM
+According to [the project site](https://llvm.org), LLVM is “a collection of modular and reusable compiler and toolchain technologies”. Basically, compilers have a lot of similar components – they need to split up the program’s control flow to define all the basic blocks, perform dead code elimination, rearrange loops, and many other optimizations. LLVM makes it easier to reuse and compose these c ...
+
+This project has an active development community and people do contribute individual optimizations and pieces of code, so architecturally this seems to have been a great choice.
```
It may also be relevant (w/r/t why to choose LLVM as a verification target) that LLVM is extremely widely _used_, so any bug in it will potentially affect a lot of compiled code.

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+This project has an active development community and people do contribute individual optimizations and pieces of code, so architecturally this seems to have been a great choice.
+
+### SMT Solvers
+SMT stands for Satisfiability Modulo Theory. Basically an SMT solver takes in a mathematical formula with many unknowns, and tries to figure out if the formula is satisfiable in the sense that there is some way to assign the variables such that the formula is true.
```
Maybe it would be worth, even very briefly, explaining how SMT is different from SAT? This description (and the example below) seems like it could apply just as much to plain old SAT.

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+Fascinatingly, despite the fact that SMT is NP-complete, computer scientists have been developing faster and faster SMT solvers, and using them in a variety of problems. While in general solving satisfiability problems requires brute force, in practice it appears that lovingly crafted (and ruthlessly empirically tested) heuristics are able to get better and better results on the problems that peo ...
+
+### Peephole Optimizations
+Peephole optimizations are small optimizations which look at a small number of instructions to find ways to do the same task but make them more computationally efficient. 
```
Maybe one example would be helpful here. Like, `-x < 0` can be rewritten to `x > 0` or whatever.

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+
+Given the prevalence of security vulnerabilities caused by improper bounds checking in array accesses such as [Heartbleed](https://en.wikipedia.org/wiki/Heartbleed), why do C specifications let you have undefined behavior rather than just defining that the behavior should be a program crash?
+
+Ralf Jung has [a good discussion](https://blog.sigplan.org/2021/11/18/undefined-behavior-deserves-a-better-reputation/), but there was also a simple (and reasonably convincing to me) example from the class discussion. When people are writing in C, they are much more likely to care a lot about performance. If the programmer knows that they didn’t mess up their loop to iterate through all of the el ...
```
Nice work capturing something from the class discussion, which would otherwise be ephemeral and lost to time! :smiley:

atucker replied:
Thanks! I learned a lot from this discussion so I wanted to make sure it made it into the blog post.

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+A core engine of the paper is in creating a domain specific language which makes it easy to write peephole optimizations which can be formally verified.
+
+#### Basic Overview
+An LLVM instruction can have source preconditions $\phi$ and target postconditions $\bar \phi$, source definedness constraints $\delta$, a source poison-free constraint $\rho$, a source result $\iota$, as well as target definedness, poison-free constraint, and result  $\bar delta$, $\bar \rho$, andt $\bar \iota$ respectively.
```
To use math in your post, you will need to set `latex = true` in the metadata, as described in [the README](https://github.com/sampsyo/cs6120). (Please take a look at your post as rendered by Zola to make sure everything looks as expected.)

sampsyo replied:
typo: "andt"

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+#### Basic Overview
+An LLVM instruction can have source preconditions $\phi$ and target postconditions $\bar \phi$, source definedness constraints $\delta$, a source poison-free constraint $\rho$, a source result $\iota$, as well as target definedness, poison-free constraint, and result  $\bar delta$, $\bar \rho$, andt $\bar \iota$ respectively.
+
+*Preconditions* can be defined using a must-analysis for precise constraints, and may-analysis for over/underapproximations. For example, the precise `isPowerOfTwo(%a)` constraint is defined as $p \leftarrow a \neq 0 \wedge a &(a-1) = 0$, while the approximate `mayAlias(%a, %b)` 
```
Not quite sure what you mean by the $\leftarrow$ symbol… is this implication? If so, maybe you want a rightward-facing arrow of some sort. It also looks like the sentence got cut off here.

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+
+*Preconditions* can be defined using a must-analysis for precise constraints, and may-analysis for over/underapproximations. For example, the precise `isPowerOfTwo(%a)` constraint is defined as $p \leftarrow a \neq 0 \wedge a &(a-1) = 0$, while the approximate `mayAlias(%a, %b)` 
+
+*Definedness* simply describes when a variable has a defined output. For example, unsigned division of a/b is defined whenever $b \neq 0$.
```
```suggestion
*Definedness* simply describes when a variable has a defined output. For example, unsigned division of `a/b` is defined whenever $b \neq 0$.
```

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+
+*Definedness* simply describes when a variable has a defined output. For example, unsigned division of a/b is defined whenever $b \neq 0$.
+
+*Poison-free* describes when a variable will not generate a poison value, which is a value which could generate undefined behavior under some circumstances. For example, left shift with no sign overflow is poison-free as long as it does not generate a sign overflow, namely as long as $(a << b) >> b = a.
```
(1) Looks like the math range wasn't closed here. (2) `<<` and `>>` probably look weird in LaTeX, so maybe you mean to use `code` instead of $math$.

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+#### Attribute Inference
+While LLVM allows developers to annotate instructions with attributes which can help enable optimizations, it is often tricky as a developer to know whether or not an attribute is appropriate. The authors also added a tool to automatically add attributes by trying to verify that the optimization with and without various attributes, aiming for the weakest preconditions and strongest postconditions ...
+
+Note that pre/postcondition here refers only to attribute annotation (like no signed overflow), not conditions for when an optimization can be applied (like isPowerOfTwo). This means that it is only used in generating the poison-free constraints.
```
```suggestion
Note that pre/postcondition here refers only to attribute annotation (like no signed overflow), not conditions for when an optimization can be applied (like `isPowerOfTwo`). This means that it is only used in generating the poison-free constraints.
```

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+
+<p align="center">
+<img src="paper_snippet.png" alt="A snippet from the paper of section 3.1.2, which describes the correctness criteria" title="Section 3.1.2 Snippet" style="zoom:25%;" />
+</p>
```
Maybe a brief, very informal textual summary of what the paper is saying here would help put it into context?

atucker replied:
Ah, good idea!

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+While many of the great points from class have already been discussed, there are a few other points that didn’t really make it into the blog post so far.
+
+### Limitations
+The paper has a few limitations which the authors also mention, such as not being able to translate all of InstCombine because it doesn't handle vectors, call/return instructions, or floating point operation.
```
Might be nice to link to [the documentation for InstCombine](https://llvm.org/docs/Passes.html#instcombine-combine-redundant-instructions)?

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+More deeply, its reliance on an SMT solver means that as more variables, operations, or bits get added to an optimization, the harder it is to verify the optimization, without there really being any way to fix the method. The paper already has issues with the handling larger bitwidths quickly for multiplication and division, and it seems that it would be hard to scale this method to handle larger ...
+
+### Defining Undefined for Fun and Profit
+One possibly-confusing thing in the paper is that the SMT solver can choose any value that it finds convenient for any `undef` value. While a glib justification may simply be that LLVM and C say that anything can happen, I think this is actually a quite reasonable and good thing for Alive to do. By choosing convenient values for `undef`, it maximizes its ability to verify specifications which agr ...
```
Not that you have to change anything here, but TBH I personally don't see any _other_ good way to deal with `undef` when modeling its semantics in SMT. Just picking zero deterministically would at least be correct, for example, but it doesn't seem to obey the spirit of verifying optimizations in the presence of `undef`: that is, it doesn't seem to help prove optimizations incorrect when they do something wrong with `undef`.

sampsyo commented on content/blog/2022-03-08-alive/index.md:
```diff
@@ -0,0 +1,114 @@
+> Beware of bugs in the above code; I have only proved it correct, not tried it.
+
+### Really? We’re just going to trust the programmer to not write code with undefined behavior?
+Finally, to return to footnote 1, it turns out that there are more recent tools that kind of let you have your cake and eat it too with regards to the performance benefits of undefined behavior. By compiling with an LLVM tool called [UndefinedBehaviorSanitizer](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html) automatically adds checks for a bunch of different types of undefined behavi ...
```
The sentence beginning with "By compiling" seems to be missing a verb in the middle. Or maybe the "By" is incorrect?

atucker replied:
Ah yeah, this got garbled.

sampsyo reviewed (changes requested):
Really nice work; this looks great! I've made a few local suggestions within.

atucker committed:
Fix headers

Fix title metadata, remove Blog from title, remove citation and linkify authors.

atucker committed:
Fix SIGPLAN -> PLDI

atucker committed:
Update content/blog/2022-03-08-alive/index.md

Co-authored-by: Adrian Sampson <adrian@radbox.org>

atucker committed:
Update index.md

Miscellaneous edits for clarity.

atucker committed:
Update index.md

Fixing based on Zola render

atucker committed:
Update index.md

Small tweaks to footnotes, and fixing a latex typo

atucker committed:
Fix last latex issue

atucker committed:
Make the paper title more different in the cite

atucker committed:
Fix footnotes.

sampsyo reviewed (approved):
Looks great! Nice work!!!
